### PR TITLE
registered execution plans

### PR DESCRIPTION
**Pull request by DvirDukhan**

### CODE REVIEW

File: src/execution_plan.c [hunk 1]
```diff
@@ -358,6 +358,10 @@ static int FlatExecutionPlan_SerializeStep(FlatExecutionStep* step, Gears_Buffer
     return FlatExecutionPlan_SerializeStepArg(type, step->bStep.arg.stepArg, bw, err);
 }
 
+static inline void FlatExecutionPlan_SerializeID(FlatExecutionPlan* fep, Gears_BufferWriter* bw) {
+    RedisGears_BWWriteBuffer(&bw, fep->id, ID_LEN);
```
Comment (MeirShpilraien) [inline src/execution_plan.c:362]: bw not &bw

File: src/execution_plan.c [hunk 2]
```diff
@@ -443,6 +447,16 @@ static const char* FlatExecutionPlan_SerializeInternal(FlatExecutionPlan* fep, s
 }
 
 int FlatExecutionPlan_Serialize(Gears_BufferWriter* bw, FlatExecutionPlan* fep, char** err){
+
```
Comment (MeirShpilraien) [inline src/execution_plan.c:450]: You need to put it in the FlatExecutionPlan_Distributed, this function is also used by rdb save and on rdb save we do want to save the flat execution ..

File: src/execution_plan.c [hunk 3]
```diff
@@ -598,8 +617,16 @@ static int FlatExecutionPlan_DeserializeInternal(FlatExecutionPlan* ret, const c
 }
 
 FlatExecutionPlan* FlatExecutionPlan_Deserialize(Gears_BufferReader* br, char** err, int encver){
-    FlatExecutionPlan* ret = FlatExecutionPlan_New();
+    bool registered = RedisGears_BRReadLong(br);
```
Comment (MeirShpilraien) [inline src/execution_plan.c:620]: Should be done on ExecutionPlan_OnReceived

File: src/execution_plan.c [hunk 4]
```diff
@@ -598,8 +617,16 @@ static int FlatExecutionPlan_DeserializeInternal(FlatExecutionPlan* ret, const c
...
+    if(registered) {
+            char *id = FlatExecutionPlan_DeserializeID(br);
+            FlatExecutionPlan* fep =  Gears_dictFetchValue(epData.registeredFepDict, id);
```
Comment (MeirShpilraien) [inline src/execution_plan.c:623]: FlatExecutionPlan_FindId

File: src/execution_plan.c [hunk 5]
```diff
@@ -598,8 +617,16 @@ static int FlatExecutionPlan_DeserializeInternal(FlatExecutionPlan* ret, const c
...
+            assert(fep);
```
Comment (MeirShpilraien) [inline src/execution_plan.c:624]: No assert, we do not want to crash here, raise error

File: src/execution_plan.c [hunk 6]
```diff
@@ -598,8 +617,16 @@ static int FlatExecutionPlan_DeserializeInternal(FlatExecutionPlan* ret, const c
...
+            RG_FREE(id);
```
Comment (MeirShpilraien) [inline src/execution_plan.c:625]: No need to free it, its part of the buffer ...

File: src/execution_plan.c [hunk 7]
```diff
@@ -2342,6 +2369,7 @@ int FlatExecutionPlan_Register(FlatExecutionPlan* fep, ExecutionMode mode, void*
     RedisModule_Replicate(ctx, RG_INNER_REGISTER_COMMAND, "b", buff->buff, buff->size);
     RedisModule_FreeThreadSafeContext(ctx);
     Gears_BufferFree(buff);
+    FlatExecutionPlan_SetRegistered(fep);
```
Comment (MeirShpilraien) [inline src/execution_plan.c:2372]: Should be done on FlatExecutionPlan_RegisterInternal which called also on slaves (and all the shards in the cluster ..)

Comment (MeirShpilraien) [reply src/execution_plan.c:2372]: Also why you need a set function for it and not just set it, if you chose a function make if static ..

File: src/execution_plan.h [hunk 8]
```diff
@@ -260,6 +260,7 @@ typedef struct FlatExecutionPlan{
     FlatBasicStep onRegisteredStep;
     FlatBasicStep onUnpausedStep;
     long long executionMaxIdleTime;
+    bool registered;
```
Comment (MeirShpilraien) [inline src/execution_plan.h:263]: Lets make it execution flags, might be more of those in the future

Comment (DvirDukhan) [reply src/execution_plan.h:263]: ok
still keeping values as true / false

File: src/execution_plan.h [hunk 9]
```diff
@@ -260,6 +261,7 @@ typedef struct FlatExecutionPlan{
     FlatBasicStep onRegisteredStep;
     FlatBasicStep onUnpausedStep;
     long long executionMaxIdleTime;
+    ExecutionFlags flags;
```
Comment (MeirShpilraien) [inline src/execution_plan.h:264]: ExecutionFlags are for executions, lets create a FlatExecutionFlags

File: src/redisgears_python.c [hunk 10]
```diff
@@ -1983,7 +1983,7 @@ static PyObject* createTensorFromBlob(PyObject *cls, PyObject *args){
     bool buffered = false;
     Py_buffer view;
     size_t size;
-    const char* blob;
+    char* blob;
```
Comment (MeirShpilraien) [inline src/redisgears_python.c:1986]: Why is this needed?

File: src/execution_plan.c [hunk 11]
```diff
@@ -1892,16 +1909,25 @@ static void ExecutionPlan_UnregisterExecutionReceived(RedisModuleCtx *ctx, const
...
+    char* err = NULL;
     Gears_BufferReader br;
     Gears_BufferReaderInit(&br, &buff);
-    char* err = NULL;
-    FlatExecutionPlan* fep = FlatExecutionPlan_Deserialize(&br, &err, REDISGEARS_DATATYPE_VERSION);
+    bool registered = RedisGears_BRReadLong(&br);
+    FlatExecutionPlan* fep;
+    if(registered) {
+        char *id = FlatExecutionPlan_DeserializeID(&br); 
+        fep = FlatExecutionPlan_FindId(id);
```
Comment (MeirShpilraien) [inline src/execution_plan.c:1926]: If fep == NULL set err here so it will be printed to the log message

File: src/execution_plan.c [hunk 12]
```diff
@@ -642,6 +650,15 @@ static void ExecutionPlan_Distribute(ExecutionPlan* ep){
...
+        if(EPIsFlagOn(ep->fep, EFRegistered)){
+        // Registered execution plan - serialize id and return.
+        RedisGears_BWWriteLong(&bw, 1);
+        FlatExecutionPlan_SerializeID(ep->fep, &bw);
+        return;
+    } else {
+        RedisGears_BWWriteLong(&bw, 0); // Non Registered execution plan.
+    }
+    
     int res = FlatExecutionPlan_Serialize(&bw, ep->fep, NULL);
```
Comment (MeirShpilraien) [inline src/execution_plan.c:662]: Notice that you serialize the FEP anyway

File: src/redisgears_python.c (hunk 10 above)
Comment (DvirDukhan) [reply src/redisgears_python.c:1986]: compile warning

File: src/execution_plan.h (hunk 9 above)
Comment (DvirDukhan) [reply src/execution_plan.h:264]: ok
I'm leaving `EPTurnOnFlag` or rename too?

File: src/execution_plan.c (hunk 12 above)
Comment (DvirDukhan) [reply src/execution_plan.c:662]: Do I need to serialize? 
I thought that we just serializing ID

Comment (MeirShpilraien) [reply src/execution_plan.c:662]: Yes but this line happened anyway in you code:
FlatExecutionPlan_Serialize(...

Comment (MeirShpilraien) [reply src/execution_plan.c:662]: Ohh sorry, missed the return ..

Comment (MeirShpilraien) [reply src/execution_plan.c:662]: But this is an issue, if you return you do not serialize the reader data. This is why the tests fails ..

File: src/execution_plan.c [hunk 13]
```diff
@@ -1507,17 +1524,13 @@ static void ExecutionPlan_RegisterForRun(ExecutionPlan* ep){
 
 void FlatExecutionPlan_AddToRegisterDict(FlatExecutionPlan* fep){
     Gears_dictAdd(epData.registeredFepDict, fep->id, fep);
-    // call the on registered callback if set
-    if(fep->onRegisteredStep.stepName){
```
Comment (MeirShpilraien) [inline src/execution_plan.c:1511]: Why you remove this code?

File: src/execution_plan.c [hunk 14]
```diff
@@ -1529,6 +1542,12 @@ static int FlatExecutionPlan_RegisterInternal(FlatExecutionPlan* fep, RedisGears
     // the registeredFepDict holds a weak pointer to the fep struct. It does not increase
     // the refcount and will be remove when the fep will be unregistered
     FlatExecutionPlan_AddToRegisterDict(fep);
+    // call the on registered callback if set
```
Comment (MeirShpilraien) [inline src/execution_plan.c:1545]: Why you returned this code here?

File: src/execution_plan.c [hunk 15]
```diff
@@ -1507,17 +1524,13 @@ static void ExecutionPlan_RegisterForRun(ExecutionPlan* ep){
 
 void FlatExecutionPlan_AddToRegisterDict(FlatExecutionPlan* fep){
     Gears_dictAdd(epData.registeredFepDict, fep->id, fep);
-    // call the on registered callback if set
-    if(fep->onRegisteredStep.stepName){
-        RedisGears_FlatExecutionOnRegisteredCallback onRegistered = FlatExecutionOnRegisteredsMgmt_Get(fep->onRegisteredStep.stepName);
-        RedisModule_Assert(onRegistered);
-        onRegistered(fep, fep->onRegisteredStep.arg.stepArg);
-    }
+    EPTurnOnFlag(fep, FEFRegistered);
```
Comment (MeirShpilraien) [inline src/execution_plan.c:1527]: This macro is for ExecutionPlan, We should create one from FlatExecutionPlan or change the name (remove the EP prefix)

File: src/execution_plan.c (hunk 13 above)
Comment (DvirDukhan) [reply src/execution_plan.c:1511]: bad rebase?

File: src/execution_plan.c (hunk 14 above)
Comment (DvirDukhan) [reply src/execution_plan.c:1545]: bad rebase?

File: src/execution_plan.c (hunk 15 above)
Comment (DvirDukhan) [reply src/execution_plan.c:1527]: added ones for FEP - copy paste for now, but if we need changes it is nice to have

File: docs/commands.md [hunk 16]
```diff
@@ -152,6 +152,7 @@ An array with an entry per execution. Each entry is made of alternating key name
 
 * **executionId**: the [execution ID](functions.md#execution-id)
 * **status**: the [status](functions.md#execution-status)
+* **registered execution**: Indication that the given execution is a registered execution plan.
```
Comment (MeirShpilraien) [inline docs/commands.md:155]: An indication that the given execution was triggered by a registration?

@itamarhaber what do you think?

Comment (itamarhaber) [reply docs/commands.md:155]: 1. I don't like the field name having a space and being so long, would replace it with `registered`
2. (note the lowercase i and no ending period)
```suggestion
* **registered**: indicates whether this is a [registered execution](functions.html#registration)
```

**Comment by MeirShpilraien** [comment]
@DvirDukhan did you run a full valgrind tests?
